flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/AbstractFineGrainedSlotManagerITCase.java

```diff
@@ -0,0 +1,672 @@
+
+        final ResourceID resourceId = ResourceID.generate();
+        final TaskExecutorConnection taskManagerConnection =
+                new TaskExecutorConnection(resourceId, taskExecutorGateway);
+        final SlotReport slotReport = new SlotReport();
+
+        new Context() {
+            {
+                mainThreadExecutor = executor;
+                runTest(
+                        () -> {
+                            getSlotManager()
+                                    .registerTaskManager(
+                                            taskManagerConnection,
+                                            slotReport,
+                                            getDefaultTaskManagerResourceProfile(),
+                                            getDefaultSlotResourceProfile());
+
+                            getSlotManager().processResourceRequirements(resourceRequirements);
+
+                            final AllocationID firstAllocationId = allocationIds.take();
+                            assertThat(allocationIds, is(empty()));
+
+                            // let the first attempt fail --> this should trigger a second attempt
+                            slotRequestFuture1.completeExceptionally(
+                                    new SlotAllocationException("Test exception."));
+
+                            executor.triggerAll();
+                            assertThat(
+                                    getTotalResourceCount(
+                                            getResourceTracker().getAcquiredResources(jobId)),
+                                    is(1));
+
+                            slotRequestFuture2.complete(Acknowledge.get());
+                            final AllocationID secondAllocationId = allocationIds.take();
+                            assertThat(allocationIds, is(empty()));
+
+                            executor.triggerAll();
+                            final TaskManagerSlotInformation slot =
+                                    getTaskManagerTracker()
+                                            .getAllocatedOrPendingSlot(secondAllocationId)
+                                            .get();
+
+                            assertThat(slot.getState(), is(SlotState.ALLOCATED));
+                            assertEquals(jobId, slot.getJobId());
+
+                            assertFalse(
+                                    getTaskManagerTracker()
+                                            .getAllocatedOrPendingSlot(firstAllocationId)
+                                            .isPresent());
+                        });
+            }
+        };
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    // Allocation update
+    // ---------------------------------------------------------------------------------------------
+
+    @Test
+    public void testAllocationUpdatesIgnoredIfTaskExecutorUnregistered() throws Exception {
```

Comment:
It's a bit unclear what this test case verifies.

IIUC, this is to verify that the system does not breakdown when the task manager is unregistered before the ack of request slot is received? I think this needs to be explained in javadoc.